### PR TITLE
New Rule T0016: Separate declarations by an empty line

### LINKED ISSUE #9130: New Rule T0016: Separate members by an empty line
Boundary between different member types should be an empty line (like between last field and first property).

These non-abstract members should be separated by empty line:
* Enums
* Multi-line fields
* Multi-line properties
* Multi-line events
* Constructors
* Destructor
* Methods
* Conversion operators
* Operators
* Nested types

It does not apply to:
* Constant fields (syntax-wise, it's a single-line field)
* Single-line fields
* Single-line properties
* Single-line events
* Abstract members

* Event fields

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #9130

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.CSharp.Styling.Test/TestCases/SeparateDeclarations.cs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        None,
+        Any,
+        All
+    }
+    private enum Enum3 {  None } // Noncompliant
+    private readonly object field1 = new();         // Noncompliant
+    private static readonly object field2 = new();
+    private readonly object field3 = new();
+    private readonly object field4, field5;
+    public abstract int AbstractMethod1();          // FN
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.CSharp.Styling.Test/TestCases/SeparateDeclarations.cs:111]: The abstract is too complicated to get right => not worth the effort.

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            || firstToken.GetPreviousToken().IsKind(SyntaxKind.CloseBraceToken)
+            || PreviousDeclarationKind(context.Node) != context.Node.Kind())
+        {
+            ValidateSeparatedMember(context);
+        }
+
+        SyntaxKind PreviousDeclarationKind(SyntaxNode node) =>
+            context.Node.Parent.ChildNodes().TakeWhile(x => x != node).LastOrDefault() is { } preceding
+                ? preceding.Kind()
+                : SyntaxKind.None;
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs:66]:
```suggestion
        static SyntaxKind PreviousDeclarationKind(SyntaxNode node) =>
            node.Parent.ChildNodes().TakeWhile(x => x != node).LastOrDefault() is { } preceding
                ? preceding.Kind()
                : SyntaxKind.None;
```

File: analyzers/tests/SonarAnalyzer.CSharp.Styling.Test/TestCases/SeparateDeclarations.cs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    public class SingleMultiLine1 { }       // Noncompliant@-3
+    /* And multiline on a single line */
+    public class SingleMultiLine2 { }       // Noncompliant@-1
+    /* And multiline on a single line */
+    /* Alaso if there are multiple */
+    public class MultipleMultiLine { }      // Noncompliant@-2
+    /// <summary>
+    /// There still should be an empty line before the documentation block
+    /// </summary>
+    public class Documentation { }          // Noncompliant@-3
```
Comment (sebastien-marichal) [inline analyzers/tests/SonarAnalyzer.CSharp.Styling.Test/TestCases/SeparateDeclarations.cs:193]: We are missing a `MultiLineDocumentationCommentTrivia` case:

```cs
    /** 
     * <summary>
     * There still should be an empty line before the documentation block
     * </summary>
     */
    public class MultilineDocumentation { }          // Noncompliant@-5
```

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                if (previousLine != trivium.GetLocation().StartLine())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                previousLine = trivium.GetLocation().EndLine();
+            }
+        }
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs:104]: I don't quite understand how the previous trivia end line could be the same as the next trivia start line.
I am not familiar with SyntaxTrivia.

**Review by sebastien-marichal — CHANGES_REQUESTED** [review]
I must admit, it took me a while to review 😅 
There is a missing test case.

There is one thing that I didn't quite understand, I left a comment.

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs:66]: The other way around - it was a leftover from other refactoring (before I declared the abstract one a FN). 

There's no reason to make it static just to pass information that is already in scope.

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/SeparateDeclarations.cs:104]: How Roslyn produces these is a solid mess and it behaves quite counter-intuitively.
```
// Line1 
^^^^^^^ This is SingleLineCommentTrivia
        ^ Here is a hidden EndOfLineCommentTrivia with the same line number

^This empty linehas also EndOfLineCommentTrivia that we're looking for
// Line 2
void Method()

/// <summary>
/// Something
/// </summary>
^^^^^^^^^^^^^^ This is a multiline SingleLineCommentTrivia that is NOT followed by EndOfLineCommentTrivia
void Method()

```

**Review by sebastien-marichal — APPROVED** [review]
LGTM!

**Pull request merged**